clippy_lints/src/dereference.rs

```diff
@@ -920,6 +944,201 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+        if !check_referent(referent) {
+            break;
+        }
+        expr = referent;
+        needless_borrow = true;
+    }
+
+    if needless_borrow {
+        Position::ImplArg(expr.hir_id)
+    } else {
+        Position::Other(precedence)
+    }
+}
+
+fn has_ref_mut_self_method(cx: &LateContext<'_>, trait_def_id: DefId) -> bool {
+    cx.tcx
+        .associated_items(trait_def_id)
+        .in_definition_order()
+        .any(|assoc_item| {
+            if assoc_item.fn_has_self_parameter {
+                let self_ty = cx.tcx.fn_sig(assoc_item.def_id).skip_binder().inputs()[0];
+                matches!(self_ty.kind(), ty::Ref(_, _, Mutability::Mut))
+            } else {
+                false
+            }
+        })
+}
+
+// Iteratively replaces `param_ty` with `new_ty` in `substs`, and similarly for each resulting
+// projected type that is a type parameter. Returns `false` if replacing the types would have an
+// effect on the function signature beyond substituting `new_ty` for `param_ty`.
+// See: https://github.com/rust-lang/rust-clippy/pull/9136#discussion_r927212757
+fn replace_types<'tcx>(
+    cx: &LateContext<'tcx>,
+    param_ty: ParamTy,
+    new_ty: Ty<'tcx>,
+    fn_sig: FnSig<'tcx>,
+    arg_index: usize,
+    projection_predicates: &[ProjectionPredicate<'tcx>],
+    substs: &mut [ty::GenericArg<'tcx>],
+) -> bool {
+    let mut replaced = BitSet::new_empty(substs.len());
+    let mut deque = VecDeque::with_capacity(substs.len());
+
+    if fn_sig
+        .inputs_and_output
+        .iter()
+        .enumerate()
+        .any(|(i, ty)| i != arg_index && contains_ty(ty, param_ty.to_ty(cx.tcx)))
+    {
+        return false;
+    }
+
+    substs[param_ty.index as usize] = ty::GenericArg::from(new_ty);
+
+    deque.push_back((param_ty, new_ty));
+
+    while let Some((param_ty, new_ty)) = deque.pop_front() {
+        for projection_predicate in projection_predicates {
+            if projection_predicate.projection_ty.self_ty() == param_ty.to_ty(cx.tcx)
+                && let ty::Term::Ty(term_ty) = projection_predicate.term
+                && let ty::Param(term_param_ty) = term_ty.kind()
+            {
+                let item_def_id = projection_predicate.projection_ty.item_def_id;
+                let assoc_item = cx.tcx.associated_item(item_def_id);
+                let projection = cx.tcx
+                    .mk_projection(assoc_item.def_id, cx.tcx.mk_substs_trait(new_ty, &[]));
+                let projected_ty = cx.tcx.normalize_erasing_regions(cx.param_env, projection);
+
+                let inserted = replaced.insert(term_param_ty.index);
+
+                let new_generic_arg = ty::GenericArg::from(projected_ty);
+
+                if substs[term_param_ty.index as usize] == new_generic_arg {
+                    continue;
+                }
+
+                // The next line provides some protection against infinite loops.
+                assert!(inserted);
```

Comment:
This can still trigger, though the case is much more contrived and requires explicitly passing a generic argument. You can do:
```rust
if insert {
    // All the work in here
}
```
The only way you can fail to insert is if the type has already been placed in the queue. If this happens the inference has already been propagated, or it will be in a future iteration. In both cases nothing needs to happen.